### PR TITLE
#5 - tsx runtime to invoke source code

### PR DESCRIPTION
**Pull request by PhantomCracker**
Resolve #7

### CODE REVIEW

**Comment by PhantomCracker** [comment]
@rndquu please check this and let me know if it works on prod: https://stackoverflow.com/questions/69220042/npm-err-unsupported-url-type-workspace-workspace

**Comment by 0x4007** [comment]
> @rndquu please check this and let me know if it works on prod: https://stackoverflow.com/questions/69220042/npm-err-unsupported-url-type-workspace-workspace

The first thing the link says is to use yarn so why dont you try that?

Also can you try using a shell script? It still makes zero sense to me that you can't use a shell script directly.

**Comment by rndquu** [comment]
> @rndquu please check this and let me know if it works on prod: https://stackoverflow.com/questions/69220042/npm-err-unsupported-url-type-workspace-workspace

Did you try how it works in your forked repo?

**Comment by PhantomCracker** [comment]
> > @rndquu please check this and let me know if it works on prod: https://stackoverflow.com/questions/69220042/npm-err-unsupported-url-type-workspace-workspace
> 
> Did you try how it works in your forked repo?

These are the tests on my side: https://github.com/PhantomCracker/action-conventional-commits/actions/runs/5689595690/job/15421394620
Please notice that on my side was working with npm too

**Comment by rndquu** [comment]
> > > @rndquu please check this and let me know if it works on prod: https://stackoverflow.com/questions/69220042/npm-err-unsupported-url-type-workspace-workspace
> > 
> > 
> > Did you try how it works in your forked repo?
> 
> These are the tests on my side: https://github.com/PhantomCracker/action-conventional-commits/actions/runs/5689595690/job/15421394620 Please notice that on my side was working with npm too

I'm getting this [error](https://github.com/rndquu/ubiquity-dollar/actions/runs/5690394676/job/15423644958)

Your latest conventional commits CI run in the `master` branch [throws](https://github.com/PhantomCracker/action-conventional-commits/actions/runs/5689730056/job/15421778053) the same error

Pls fix the `master` branch

**Comment by PhantomCracker** [comment]
Yes, that was because I was trying to do something with SH file, but it looks like the emacs doesn't know where and what the yarn is :)
here is the latest test: https://github.com/PhantomCracker/action-conventional-commits/actions/runs/5690465733

**Comment by rndquu** [comment]
> Yes, that was because I was trying to do something with SH file, but it looks like the emacs doesn't know where and what the yarn is :) here is the latest test: https://github.com/PhantomCracker/action-conventional-commits/actions/runs/5690465733

I don't really understand why it works in your repo but [fails](https://github.com/rndquu/ubiquity-dollar/actions/runs/5690394676/job/15424158179) in my

**Comment by PhantomCracker** [comment]
> fails

Ok, good question, let me investigate this and make a fix for it
Be right back with an answer

**Comment by PhantomCracker** [comment]
Hei @rndquu ,
I was checking what you are testing and now it's clear for me.
If I run the action on the action-conventional-commits repository everything is fine as you can see here: https://github.com/PhantomCracker/action-conventional-commits/actions/runs/5699435599/job/15448346012

BUT, if I run the action in another repository (for example like you did it on ubiquity-dollar), we receive the following error that says that the src/main.ts file does not exist:
Screenshot 2023-07-29 at 10 55 09

So what we need to do is to run all the files in the context of action-conventional-commits repository; to do this we need to run the dist/index.js file :) 
If you have any other idea, please let me know, I will also try new approaches.

**Comment by rndquu** [comment]
> So what we need to do is to run all the files in the context of action-conventional-commits repository

I don't understand what "run all the files in the context of action-conventional-commits repository" means. Did you manage to run the conventional commits workflow in a 3rd party repo, not in the https://github.com/PhantomCracker/action-conventional-commits ?

**Comment by PhantomCracker** [comment]
> > So what we need to do is to run all the files in the context of action-conventional-commits repository
> 
> I don't understand what "run all the files in the context of action-conventional-commits repository" means. Did you manage to run the conventional commits workflow in a 3rd party repo, not in the https://github.com/PhantomCracker/action-conventional-commits ?

I mean that as you see in the screenshot that I have shared, the `src/main.ts` is not found in ubiquity dollar repository; basically, this happens because we run the following command in `src/index.js`:
`yes 2>/dev/null | npx tsx src/main.ts`
So we need to keep all the actions in the context of action-conventional-commits repository i think...
And I would like to know if you thinked on something else for this issue

File: .github/workflows/test.yml
Comment (rndquu) [inline .github/workflows/test.yml:1]: Pls remove this file

File: src/index.js
Comment (rndquu) [inline src/index.js:1]: Pls remove this file either along with `dist/main/index.js`

**Review by rndquu — CHANGES_REQUESTED** [review]
So github supports [JS actions](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action) but they require either using [ncc](https://github.com/vercel/ncc) to compile all dependencies in a single file either commiting the whole `node_modules` folder

There are also [composite actions](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) which allow to do pretty much everything

If we want to run `tsx` without commiting dependencies we need to: 
- use composite action instead of JS action
- in the action's yml download `action-conventional-commits`, install dependencies and run `tsx`

@PhantomCracker pls refactor [this way](https://github.com/rndquu/action-conventional-commits/blob/ci/composite/action.yml)

**Comment by PhantomCracker** [comment]
I see... thank you for explanations, pull request has been updated.

File: action.yml [hunk 1]
```diff
@@ -5,5 +5,18 @@ inputs:
...
-  using: node16
-  main: dist/main/index.js
+  using: "composite"
+  steps:
+    - name: Setup Node
+      uses: actions/setup-node@v3
+      with:
+        node-version: "18.14.1"
+    - name: Git clone action-conventional-commits repo to the home directory
+      run: cd ~ && git clone https://github.com/rndquu/action-conventional-commits
```
Comment (rndquu) [inline action.yml]:
```suggestion
      run: cd ~ && git clone https://github.com/ubiquity/action-conventional-commits
```

**Review by rndquu — CHANGES_REQUESTED** [review]

**Review by rndquu — APPROVED** [review]

**Pull request merged**